File: 02 medium/search problems/733_flood_fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} color
 * @return {number[][]}
 */
var floodFill = function(image, sr, sc, color) {

    let source = image[sr][sc]

    const dfs = (sr, sc) => {
        if (sr < 0 || sc < 0 || sr >= image.length || sc >= image[0].length || image[sr][sc] !== source)  {
            return
        }
        
        image[sr][sc] = color
        
        dfs(sr + 1, sc)   
        dfs(sr - 1, sc)   
        dfs(sr, sc + 1)   
        dfs(sr, sc - 1)      
    }

    if (source != color) dfs(sr, sc) 
    return image
};

File: 02 medium/search problems/graph search/733_flood_fill.py
class Solution(object):
    def floodFill(self, image, sr, sc, color):
        """
        :type image: List[List[int]]
        :type sr: int
        :type sc: int
        :type color: int
        :rtype: List[List[int]]
        """
        global rows, cols, og
        rows = len(image) 
        cols = len(image[0])
        og = image[sr][sc]

        def fill(sr, sc): 
            global rows, cols, og
            if sr < 0 or sr >= rows or sc < 0 or sc >= cols or image[sr][sc] != og: 
                return 
            
            image[sr][sc] = color 

            fill(sr + 1, sc) 
            fill(sr - 1, sc) 
            fill(sr, sc + 1) 
            fill(sr, sc - 1)

        if image[sr][sc] != color: 
            fill(sr, sc) 
            
        return image 